Add PAL qualification tests for v3 of the OTA library

yanjos-dev opened the pull request:
Add PAL qualification tests for v3 of the OTA library

Description
-----------
There were unit tests for v1 of the OTA library before PR https://github.com/aws/amazon-freertos/pull/3166 removed them along with the library. Before this PR removed them, there were the FULL_OTA_CBOR, FULL_OTA_AGENT, and FULL_OTA_PAL test groups.

This PR adds back in a version of the FULL_OTA_PAL test group that has been updated to be compatible with v3 of the OTA library that was introduced in PR https://github.com/aws/amazon-freertos/pull/3185.

The FULL_OTA_CBOR and FULL_OTA_AGENT test groups are not being added back in. These test groups were originally testing the library itself and were unrelated to porting devices. The OTA library now has [unit tests that exist inside of the OTA repository](https://github.com/aws/ota-for-aws-iot-embedded-sdk/tree/v3.0.0/test/unit-test) so these groups are no longer needed.

This PR was developed in a branch (feature/ota_v3_integration) that is too large to merge all at once so this PR only consists of the test related code from the development branch.

**The original commit history, review, and list of changes can be found in PR** https://github.com/aws/amazon-freertos/pull/3085.

Checklist:
----------
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] I have tested my changes. No regression in existing tests.
- [ ] My code is Linted.


By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

yanjos-dev committed:
Add OTA PAL qualification tests

yanjos-dev committed:
Remove references to deprecated OTA unit tests

* Remove references to FULL_OTA_CBOR test group.
* Remove references to FULL_OTA_AGENT test group.

yanjos-dev commented:
/bot run checks

yanjos-dev commented:
/bot run checks

divekarshubham commented on tests/integration_test/ota_pal/aws_test_ota_pal.c:
```diff
@@ -0,0 +1,1137 @@
+/**
+ * @brief Set the image state to unknown and verify a failure.
+ */
+TEST( Full_OTA_PAL, otaPal_SetPlatformImageState_UnknownImageState )
```
Most of the SetPlatformImageState tests can be combined to one:
```C
prvSetAndVerifyPlatformState(ImageState, ExpectedPalStatus)
{
    OtaPalStatus_t xOtaStatus;
    OtaImageState_t eImageState = ImageState;
    int16_t blocksWritten;
  .....
  
        xOtaStatus = otaPal_SetPlatformImageState( &xOtaFile, eImageState );
        TEST_ASSERT_EQUAL( ExpectedPalStatus, OTA_PAL_MAIN_ERR( xOtaStatus ) );
}

TEST( Full_OTA_PAL, otaPal_SetPlatformImageState_UnknownImageState )
{
/*Test for unknown state*/
prvSetAndVerifyPlatformState(OtaImageStateUnknown, OtaPalSuccess)
....
}

yanjos-dev replied:
It's realistic to have a device fail a subset of the SetPlatformImageState tests so I would like to keep these separate for now.

divekarshubham replied:
You could still keep the tests separate and have them re-use the testing code.

yanjos-dev replied:
^Resolved this offline.

yanjos-dev commented on tests/integration_test/ota_pal/README.md:
```diff
@@ -0,0 +1,19 @@
+#OTA PAL Testing
+
+Please provision your device with the correct certificates in \<root-directory\>/tests/integration_test/ota_pal/test_files before running any OTA PAL tests( these tests are enabled by setting testrunnerFULL_OTA_PAL_ENABLED to 1, see aws_test_runner_config.h ).
```
test_files is the name of the directory containing the test files. Did you mean to add <signature-file> to the end of that instead of replacing?

divekarshubham replied:
I don't see this directory in https://github.com/yanjos-dev/amazon-freertos/tree/yanjos/add_ota_pal_tests/tests/integration_test/ota_pal

yanjos-dev replied:
Thanks for pointing this out. This was previously being ignored by git because of the file path names. I have added .test to the name of the files to clearly point out that they are test files and to avoid them being ignored.

yanjos-dev replied:
The rest of the feedback with putting the tick around files and macros I have added in 2633135.

yanjos-dev commented on tests/integration_test/ota_pal/README.md:
```diff
@@ -0,0 +1,19 @@
+
+The names of the files correspond to the signing algorithm associated with that certificate.
+
+If your OTA PAL layer port defines "otapalconfigCODE_SIGNING_CERTIFICATE" variable in ota_pal.c, then please copy in the certificate, with the applicable signing algorithm for your device, located under \<root-directory\>/tests/integration_tests/ota_pal/test_files. Currently the "otapalconfigCODE_SIGNING_CERTIFICATE" variable is set to use the ecdsa-sha256-signer.crt.pem certificate located under ...
```
same as other comment, this one I think the file path isn't quite correct right?

divekarshubham replied:
All signature files seem to be a part of  tests/integration_test/ota_pal

yanjos-dev replied:
Same as previous, see 2633135.

yanjos-dev committed:
Add OTA test certs as an exception to git ignore

yanjos-dev committed:
Revert "Add OTA test certs as an exception to git ignore"

This reverts commit 78f9e4bc92bbf44eeb124662df186999c2106934.

yanjos-dev committed:
Update the name of the OTA test certificate files

* These certificates are required for testing. They were accidentally
  not being tracked by git due to the file names. This change updates
  the names to end in .test so that they are not ignored and the purpose
  of them is made more clear.

yanjos-dev committed:
Update OTA test README.md based on PR feedback

yanjos-dev committed:
Update file header for OTA test files

yanjos-dev committed:
Update OTA PAL test logging stack name

yanjos-dev commented on tests/integration_test/ota_pal/aws_test_ota_pal.c:
```diff
@@ -0,0 +1,1137 @@
+ */
+TEST( Full_OTA_PAL, otaPal_CloseFile_NonexistingCodeSignerCertificate )
+{
+    #if ( otatestpalUSE_FILE_SYSTEM == 1 )
```
This was a pre-existing test that I did not write, so I can only speculate.

My understanding is that this test is only valid if a device has a file system. This could be a normal if statement, but I think this makes it more clear that this will never run if the device does not support a file system.

yanjos-dev commented on tests/integration_test/ota_pal/aws_test_ota_pal.c:
```diff
@@ -0,0 +1,1137 @@
+        TEST_ASSERT_EQUAL_INT( OtaPalSuccess, OTA_PAL_MAIN_ERR( xOtaStatus ) );
+    }
+}
+#if ( otatestpalREAD_CERTIFICATE_FROM_NVM_WITH_PKCS11 == 1 )
```
What is the reasoning with this? I believe this is where it is so that the definition is directly above where it is used.

yanjos-dev commented on tests/integration_test/ota_pal/aws_test_ota_pal.c:
```diff
@@ -0,0 +1,1137 @@
+
+    /* We use a dummy file name here because closing the system designated bootable
+     * image with content that is not runnable may cause issues. */
+    xOtaFile.pFilePath = ( uint8_t * ) ( "test_happy_path_image.bin" );
+    xOtaFile.fileSize = sizeof( ucDummyData );
+    xOtaStatus = otaPal_CreateFileForRx( &xOtaFile );
+    TEST_ASSERT_EQUAL( OtaPalSuccess, OTA_PAL_MAIN_ERR( xOtaStatus ) );
```
This is only used in two places so this won't reduce the amount of code. IMO we should keep this as is so we aren't introducing a layer of abstraction.

yanjos-dev committed:
Remove commented out code

yanjos-dev commented on tests/integration_test/ota_pal/README.md:
```diff
@@ -0,0 +1,19 @@
+#OTA PAL Testing
```
Done.

yanjos-dev commented on tests/integration_test/ota_pal/aws_ota_pal_test_access_declare.h:
```diff
@@ -0,0 +1,41 @@
+ *
+ * http://aws.amazon.com/freertos
+ * http://www.FreeRTOS.org
+ */
```
Done.

yanjos-dev commented on tests/integration_test/ota_pal/aws_test_ota_pal.c:
```diff
@@ -0,0 +1,1137 @@
+
+/* Logging configuration for the Demo. */
+#ifndef LIBRARY_LOG_NAME
+    #define LIBRARY_LOG_NAME    "TEST"
```
Good idea, done.

yanjos-dev commented on tests/integration_test/ota_pal/aws_test_ota_pal.c:
```diff
@@ -0,0 +1,1137 @@
+ *
+ * http://aws.amazon.com/freertos
+ * http://www.FreeRTOS.org
+ */
```
Done.

yanjos-dev replied:
Nice catch, I missed this when I was adding headers to the other files. It should be updated now. 8bbc1af

yanjos-dev commented on tests/integration_test/ota_pal/aws_test_ota_pal.c:
```diff
@@ -0,0 +1,1137 @@
+         */
+
+        /*ePalImageState = otaPal_GetPlatformImageState();
+         * TEST_ASSERT_EQUAL_INT( OtaPalImageStatePendingCommit, ePalImageState ); */
```
I think this was originally left there to show the test that we are ignoring because of device limitations. But I agree this could be confusing so I have removed this. There is still a comment above here explaining why we aren't doing that specific check.

yanjos-dev committed:
Merge main into yanjos/add_ota_pal_tests

aggarg commented on tests/integration_test/ota_pal/aws_test_ota_pal_ecdsa_sha256_signature.h:
```diff
@@ -0,0 +1,69 @@
+ * http://www.FreeRTOS.org
+ */
+
+#ifndef _AWS_TEST_PAL_ECDSA_SHA256_SIGNATURE_H_
```
```suggestion
#ifndef AWS_TEST_PAL_ECDSA_SHA256_SIGNATURE_H_
```

yanjos-dev replied:
Done

aggarg commented on tests/integration_test/ota_pal/aws_test_ota_pal_rsa_sha1_signature.h:
```diff
@@ -0,0 +1,117 @@
+ * http://www.FreeRTOS.org
+ */
+
+#ifndef _AWS_TEST_OTA_PAL_RSA_SHA1_SIGNATURE_H_
```
```suggestion
#ifndef AWS_TEST_OTA_PAL_RSA_SHA1_SIGNATURE_H_
```

yanjos-dev replied:
Done

aggarg commented on tests/integration_test/ota_pal/aws_test_ota_pal_rsa_sha256_signature.h:
```diff
@@ -0,0 +1,114 @@
+ * http://www.FreeRTOS.org
+ */
+
+#ifndef _AWS_TEST_OTA_PAL_RSA_SHA256_SIGNATURE_H_
```
This applies to all the places. Names starting with underscore are reserved as per the spec and therefore we should not use them. 
```suggestion
#ifndef AWS_TEST_OTA_PAL_RSA_SHA256_SIGNATURE_H_
```

yanjos-dev replied:
Done

yanjos-dev committed:
Add a file header to aws_test_ota_pal.c

yanjos-dev committed:
Update the OTA PAL test README.md

yanjos-dev commented on tests/integration_test/ota_pal/README.md:
```diff
@@ -0,0 +1,19 @@
+
+The names of the files correspond to the signing algorithm associated with that certificate.
+
+If your OTA PAL layer port defines `otapalconfigCODE_SIGNING_CERTIFICATE` variable in `ota_pal.c`, then please copy in the certificate, with the applicable signing algorithm for your device, located under `<root-directory>/tests/integration_test/ota_pal/test_files/`. The `otapalconfigCODE_SIGNING_CERTIFICATE` variable is set to use the `<root-directory>/tests/integration_test/ota_pal/test_files/e ...
```
I have updated this README.md file to make it more clear in general. I think that the original description does not make as much since after updating the tests. Can you look over this?

yanjos-dev commented on tests/integration_test/ota_pal/aws_test_ota_pal.c:
```diff
@@ -0,0 +1,1134 @@
+
+    if( TEST_PROTECT() )
+    {
+        blocksWritten = otaPal_WriteBlock( &xOtaFile, 0, &ucData, 1 );
```
This is true, for these write block tests and some of the other ones.

Ideally this test would do some kind of read and validate that the blocks written match what was read back. We don't have an API that is implemented by each platform for reading from a section in the storage mechanism so we can't do this kind of functionality.

There is some validation that it is actually writing a block in other tests though. The close file functions that verify the signature will accomplish this. I expect those tests to fail without successful writes. This is also softly validated by the end-to-end OTA process that is part of the qualification process.

I do think these tests have a place to give some surface level insight on the return values of the function so I'd still like to keep them despite them not doing a full validation.

yanjos-dev committed:
Update include guards of OTA test files

yanjos-dev commented on tests/integration_test/ota_pal/aws_ota_pal_test_access_declare.h:
```diff
@@ -0,0 +1,48 @@
+ * Needed for testing private functions.
+ */
+
+#ifndef _AWS_OTA_PAL_TEST_ACCESS_DECLARE_H_
```
Done.

yanjos-dev committed:
Update tests/integration_test/ota_pal/aws_test_ota_pal.c

Co-authored-by: Gaurav-Aggarwal-AWS <33462878+aggarg@users.noreply.github.com>

divekarshubham commented:
/bot run checks

nateglims commented:
/bot run checks

yanjos-dev committed:
Empty commit to restart the build checks
